File: solutions/ruby/yacht/1/yacht.rb
class Yacht
    RULES = {
        'yacht' => lambda { |dice| dice.uniq.length == 1 ? 50 : 0 },
        'little straight' => lambda { |dice| dice.sort == [1, 2, 3, 4, 5] ? 30 : 0 },
        'big straight' => lambda { |dice| dice.sort == [2, 3, 4, 5, 6] ? 30 : 0 },
        'full house' => lambda { |dice| score_full_house(dice) },
        'four of a kind' => lambda { |dice| score_four_of_a_kind(dice) },
        'choice' => lambda { |dice| dice.sum },
        'ones' => lambda { |dice| score_kind(dice, 1) },
        'twos' => lambda { |dice| score_kind(dice, 2) },
        'threes' => lambda { |dice| score_kind(dice, 3) },
        'fours' => lambda { |dice| score_kind(dice, 4) },
        'fives' => lambda { |dice| score_kind(dice, 5) },
        'sixes' => lambda { |dice| score_kind(dice, 6) }
    }

    def initialize(dice, type)
        @dice = dice
        @type = type
    end

    def score
        if RULES.has_key?(@type)
            RULES[@type].call(@dice)
        else
            0
        end
    end

    private

    def self.score_full_house(dice)
        return 0 if dice.uniq.length != 2
        return 0 if dice.count(dice[0]) != 2 && dice.count(dice[0]) != 3
        dice.sum
    end

    def self.score_four_of_a_kind(dice)
        return dice[0] * 4 if dice.count(dice[0]) >= 4
        return dice[1] * 4 if dice.count(dice[1]) >= 4
        0
    end

    def self.score_kind(dice, kind)
        dice.count(kind) > 0 ? dice.count(kind) * kind : 0
    end
end

File: solutions/javascript/yacht/1/yacht.js
const rules = {
  ones: (dice) => dice.filter((d) => d === 1).length,
  twos: (dice) => dice.filter((d) => d === 2).length * 2,
  threes: (dice) => dice.filter((d) => d === 3).length * 3,
  fours: (dice) => dice.filter((d) => d === 4).length * 4,
  fives: (dice) => dice.filter((d) => d === 5).length * 5,
  sixes: (dice) => dice.filter((d) => d === 6).length * 6,
  'full house': (dice) => {
    const counts = dice.reduce((acc, d) => {
      acc[d] = acc[d] + 1 || 1;
      return acc;
    }, {});
    return (Object.values(counts).sort().join('') === '23') ? dice.reduce((acc, d) => acc + d, 0) : 0;
  },
  'four of a kind': (dice) => {
    const counts = dice.reduce((acc, d) => {
      acc[d] = acc[d] + 1 || 1;
      return acc;
    }, {});
    return Object.entries(counts).reduce((acc, [k, v]) => (v >= 4) ? k * 4 : acc, 0);
  },
  'little straight': (dice) => (dice.sort().join('') === '12345') ? 30 : 0,
  'big straight': (dice) => (dice.sort().join('') === '23456') ? 30 : 0,
  choice: (dice) => dice.reduce((acc, d) => acc + d, 0),
  yacht: (dice) => (dice.every((d) => d === dice[0])) ? 50 : 0,
};

export const score = (dice, category) => {
  if (!rules[category]) {
    throw new Error('Invalid category');
  }
  return rules[category](dice);
};
